Checkers v2 - Leaderboard and Migration (#1189)
cosmos/sdk-tutorials: academy/3-my-own-chain/migration.md

---
title: "Introduce a Leaderboard After Production"
order: 20

> description: A leaderboard for your in-production blockchain

Author: ```suggestion
description: A leaderboard for your in-production blockchain via state migration
```

<HighlightBox type="tip">

> If you feel unsure about creating new data structures with Ignite CLI, look at the [previous sections](/hands-on-exercise/1-ignite-cli/create-message.md) of the exercise again.

Author: ```suggestion
If you feel unsure about creating new data structures with Ignite CLI, look at the [previous sections](/academy/3-my-own-chain/create-message.md) of the exercise again.
```

    }
    ```

> 3. Add a **leaderboard rung structure** to be repeated inside the leaderboard: this stores the information of a player scoring high enough to be included in the leaderboard. It is not meant to be kept directly in storage as it is only a part of the leaderboard. Instead of involving Ignite CLI, create the structure by hand in its file:

Author: By removing _own_, the reader will not know that this is a new file that they have to create. They will look for it in the folder structure and tell us on Discord that it does not exist.

```suggestion
3. Add a **leaderboard rung structure** to be repeated inside the leaderboard: this stores the information of a player scoring high enough to be included in the leaderboard. It is not meant to be kept directly in storage as it is only a part of the leaderboard. Instead of involving Ignite CLI, create the structure by hand in a new file:
```

Reviewer: I would recommend reverting to "in its own file".

Author: Now I kinda like the "in a new file" 🤣

Reviewer: haha, then let it be so

<ExpansionPanel title="How is dateAdded (de)serialized?">

> You can reuse the [date format used for the deadline](/hands-on-exercise/2-ignite-cli/game-deadline.md):

Author: ```suggestion
You can reuse the [date format used for the deadline](/academy/3-my-own-chain/game-deadline.md):
```

<ExpansionPanel title="How is dateAdded (de)serialized?">

> You can reuse the [date format used for the deadline](/hands-on-exercise/2-ignte-cli-adv/2-game-deadline.md):

Author: `ignte` 😬 ?

</ExpansionPanel>

> <ExpansionPanel title="Add functions to (de-serialize WinningPlayerParsed">

Author: ```suggestion
<ExpansionPanel title="Add functions to (de)serialize WinningPlayerParsed">
```

Writing _(de)action-word_ is a common way of saying that we are going to talk about both _serialize_ and _deserialize_.

> <ExpansionPanel title="Add functions to (de-serialize WinningPlayerParsed">

Author: ```suggestion
<ExpansionPanel title="Add functions to serialize and deserialize WinningPlayerParsed">
```

1. Call [`keeper.GetAllStoredGame()`](https://github.com/cosmos/b9-checkers-academy-draft/blob/migration/x/checkers/keeper/stored_game.go#L50) to get an array with all the games.
2. Keep only the games that have a winner.
3. Then for each game:

>     1. Call `keeper.GetPlayerInfo` or, if that is not found, create player info, both for the black player and the red player.

Author: ```suggestion
    1. Call `keeper.GetPlayerInfo` or, if that is not found, create a player info record, both for the black player and the red player.
```

1. The **game processing** routine will:
    
    * Receive separate arrays of games from the _stored-game_ channel.

>     * Compute the aggregate player info files from them. I.e. **_map_**.

Author: Mmh, no _files_ does not work well here. All is happening in memory and files are understood as on disk.

_Records_ or _objects_ works better.

    }

    if app.UpgradeKeeper.IsSkipHeight(upgradeInfo.Height) {
        return
    }

    var storeUpgrades *storetypes.StoreUpgrades

    switch upgradeInfo.Name {
    case v1tov2.UpgradeName:
    }

    if storeUpgrades != nil {
        // configure store loader that checks if version == upgradeHeight and applies store upgrades
        app.SetStoreLoader(upgradetypes.UpgradeStoreLoader(upgradeInfo.Height, storeUpgrades))
    }
}
```

Now you are ready to inform the app proper. You do this towards the end, after the call to `app.SetEndBlocker` and before `if loadLatest`. At the right location:

```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/migration/app/app.go#L610]
...
app.SetEndBlocker(app.EndBlocker)

app.setupUpgradeHandlers()

if loadLatest {
    ...
}
```

One more thing, for some reason, the `monitoring` module added by Ignite gives difficulty when experimenting below with the CLI. To make it simple, you remove [all references to `monitoring`](https://github.com/cosmos/b9-checkers-academy-draft/compare/leaderboard-handling..migration#diff-0f1d2976054440336a576d47a44a37b80cdf6701dd9113012bce0e3c425819b7L159) from `app.go`.

> That's it.

Author: Feel free to change. I wanted to convey the fact that:

1. creating the player info and the leaderboard is a lot of work.
2. adding the store migration functions is some work but less than 1.
3. adding the callbacks is quite short when done right.

func (app *App) setupUpgradeHandlers() {
    // v1 to v2 upgrade handler
    app.UpgradeKeeper.SetUpgradeHandler(
        v1tov2.UpgradeName,
        func(ctx sdk.Context, plan upgradetypes.Plan, vm module.VersionMap) (module.VersionMap, error) {
            return app.mm.RunMigrations(ctx, app.configurator, vm)
        },
    )

    // When a planned update height is reached, the old binary will panic
    // writing on disk the height and name of the update that triggered it
    // This will read that value, and execute the preparations for the upgrade.
    upgradeInfo, err := app.UpgradeKeeper.ReadUpgradeInfoFromDisk()
    if err != nil {
        panic(fmt.Errorf("failed to read upgrade info from disk: %w", err))
    }

    if app.UpgradeKeeper.IsSkipHeight(upgradeInfo.Height) {
        return
    }

    var storeUpgrades *storetypes.StoreUpgrades

    switch upgradeInfo.Name {
    case v1tov2.UpgradeName:
    }

    if storeUpgrades != nil {
        // configure store loader that checks if version == upgradeHeight and applies store upgrades
        app.SetStoreLoader(upgradetypes.UpgradeStoreLoader(upgradeInfo.Height, storeUpgrades))
    }
}
```

> Now you are ready to inform the app proper. You do this towards the end, after the call to `app.SetEndBlocker` and before `if loadLatest`. At the correct location:

Author: The app proper. Before it was dealing within modules, now we inform the app itself.

$ export bob=$(./release/v1/checkersd keys show bob -a)
$ ./release/v1/checkersd tx checkers create-game \
    $alice $bob 10 stake \
    --from $alice --keyring-backend test --yes \
    --broadcast-mode block
```

</CodeGroupItem>

<CodeGroupItem title="Docker">

```sh
$ export alice=$(docker exec checkers ./release/v1/checkersd keys show alice -a --keyring-backend test)
$ export bob=$(docker exec checkers ./release/v1/checkersd keys show bob -a --keyring-backend test)
$ docker exec -it checkers ./release/v1/checkersd tx checkers create-game \
    $alice $bob 10 stake \
    --from $alice --keyring-backend test --yes \
    --broadcast-mode block
```

</CodeGroupItem>

</CodeGroup>

<HighlightBox type="note">

The `--broadcast-mode block` flag means that you can fire up many such games by just copying the command without facing any sequence errors.

</HighlightBox>

To get a few complete games, you are going to run the [integration tests](https://github.com/cosmos/academy-checkers-ui/blob/server-indexing/test/integration/stored-game-action.ts) against it. These tests were built to run against a running chain started by Ignite. What is different here is that:

> 1. Blocks come slower, likely every five seconds instead of every one.

Reviewer 2: @xavierlepretre Should this be a numbered list or a bullet point one?

Author: I put numbers because the next list should say _respectively_.

    $alice $bob 10 stake \
    --from $alice --keyring-backend test --yes \
    --broadcast-mode block
```

</CodeGroupItem>

<CodeGroupItem title="Docker">

```sh
$ export alice=$(docker exec checkers ./release/v1/checkersd keys show alice -a --keyring-backend test)
$ export bob=$(docker exec checkers ./release/v1/checkersd keys show bob -a --keyring-backend test)
$ docker exec -it checkers ./release/v1/checkersd tx checkers create-game \
    $alice $bob 10 stake \
    --from $alice --keyring-backend test --yes \
    --broadcast-mode block
```

</CodeGroupItem>

</CodeGroup>

<HighlightBox type="note">

The `--broadcast-mode block` flag means that you can fire up many such games by just copying the command without facing any sequence errors.

</HighlightBox>

To get a few complete games, you are going to run the [integration tests](https://github.com/cosmos/academy-checkers-ui/blob/server-indexing/test/integration/stored-game-action.ts) against it. These tests were built to run against a running chain started by Ignite. What is different here is that:

1. Blocks come slower, likely every five seconds instead of every one.
2. There are no longer any faucets.

> Therefore, to be able to run these tests you need to:

Author: ```suggestion
Therefore, to be able to run these tests you need to attend to the above problems, respectively:
```
@CitMC

It looks beneficial to use a Go routine in this case too, and a _player info_ channel to pass along arrays of player info records.

> In practice, repeatedly building the intermediate leaderboard means adding _k_ new `winningPlayerParsed` to the sorted array, sorting it, clipping it to `LeaderboardWinnerLength`, and repeating. What constitutes a good _k_ value should be dictated by testing and performance measurements. However, you can start with your best guess in its file:

Author: ```suggestion
In practice, repeatedly building the intermediate leaderboard means adding _k_ new `winningPlayerParsed` to the sorted array, sorting it, clipping it to `LeaderboardWinnerLength`, and repeating. What constitutes a good _k_ value should be dictated by testing and performance measurements. However, you can start with your best guess in a new file created for this purpose:
```

## v1 to v2 migration proper

> The migration proper needs to execute the previous functions in a specific order. You can encapsulate this knowledge in its function:

Author: ```suggestion
The migration proper needs to execute the previous functions in a specific order. You can encapsulate this knowledge in a function:
```

}
```

<HighlightBox type="note">

This does not panic in case of an error. To avoid carrying on a faulty state, the caller of this function will have to handle the panic.

</HighlightBox>

You have in place the functions that will handle the store migration. Now you have to set up the chain of command for these functions to be called by the node at the right point in time.

### Consensus version and name

> The `upgrade` module keeps in its own store of the [different module versions](https://docs.cosmos.network/master/core/upgrade.html#tracking-module-versions) that are currently running. To signal an upgrade, your module needs to return a different value when queried by the `upgrade` module. Change it from `2` to `3`, or whichever number works for you. First keep both these values in their respective locations:

Author: No no.
It is:
The `upgrade` module keeps the different module versions. Where? In its own store.

Reviewer: Corrected.

    --rpc.laddr "tcp://0.0.0.0:26657"
```

> <HighlightBox type="note">

Author: I think that it will not show as a highlight box because it is already in a `CodeGroupItem`.

Reviewer: Corrected.
